Consolidate validation in Block and Transaction

hashmap opened the pull request:
This PR Introduces TransactionBody which is included into block and tx.

Initially I tied to hide `body` by exposing `inputs/outputs/kernels` as methods on tx and block, but then I found that exposing `body` as a filed makes it conceptually simpler, but perhaps harder to read/use.

Fixes #1333

ignopeverell commented:
Yes, I like TransactionBody like this, and it factors in a lot of validation code, which is great. Now for the bad news... I don't really like the `.body.kernels` or `.body.outputs` everywhere. It'd be a lot nicer with a `inputs()`, `outputs()` and `kernels()` accessor on `Transaction` and `Block`. May be a pain to introduce but I think it'd be a lot nicer.

hashmap commented:
Not a pain, I did one round of refactoring in the opposite direction already, thanks to mighty vim. I’ll do it one more time, unfortunately it will add some amount of `clone()`

ignopeverell commented:
Perhaps the accessors should return references then?

hashmap commented:
That’s how I started. Then found that in some places we consume them, eg in `aggregate`. Anyway now I have couple of ideas how to implement it with minimum overhead.

hashmap commented:
@ignopeverell I removed direct `body` access. In places where we consume `i/o/k` I introduced a type safe zero-cost conversion of `tx` into `tx body`, so we can access `i/o/k` fields without `clone` even if `body` is private.

antiochp commented on core/src/genesis.rs:
```diff
@@ -25,109 +25,83 @@ use global;
-		outputs: vec![],
-		kernels: vec![],
-	}
+	core::Block::with_header(core::BlockHeader {
```
Nice!

antiochp commented on core/src/core/block.rs:
```diff
@@ -47,7 +47,7 @@ pub enum Error {
 	/// Restrict block total weight.
 	TooHeavy,
-	/// Block weight (based on inputs|outputs|kernels) exceeded.
+	/// Block weight (based on.body.inputs.body.outputs.body.kernels) exceeded.
```
Comment messed up during search/replace?

antiochp commented on core/src/core/block.rs:
```diff
@@ -488,9 +452,10 @@ impl Block {
-			all_outputs.extend(tx.outputs);
-			all_kernels.extend(tx.kernels);
+			let tb: TransactionBody = tx.into();
+			all_inputs.extend(tb.inputs);
```
Most access appears to be along the pattern of `block.inputs()` or `tx.inputs()` now.
It may feel more consistent if we could do this here also (via an accessor and not a direct field access)?

hashmap replied:
Accessors borrow references to fields, here we need actual values, not references. Direct field access in Rust allows partial move of values from the struct, which is quite unique for Rust. I tried to make this case explicit by type casting, instead of returning `block` filed from the method (which implicitly destroys the tx here, quite unexpected side effect). Another approach would be clone, but I don't like it here because of loop, overhead may be quite big.

antiochp replied:
That makes sense I think. I still don't have an intuitive grasp of this in Rust... I think I get what you are saying.

antiochp commented on core/src/core/block.rs:
```diff
@@ -533,7 +497,7 @@ impl Block {
 		let mut kern_ids = vec![];
 
-		for k in &self.kernels {
+		for k in &self.body.kernels {
```
Can we use the accessor here? `&self.kernels()`

hashmap commented:
@ignopeverell sure, I run my node on it (with fast sync from scratch after each build)
